### PR TITLE
Prepare snow depth observations for JEDI-based Land DA.

### PR DESCRIPTION
**Pull request by aerorahul**
**Description**
This PR:
- adds a job to prepare IMS snow depth observations as a task in the workflow.  This task depends on the `prep.sh` job to bring IMS data.  To test this type of data `DMPDIR` in `config.base` needed to be pointed to `"/scratch1/NCEPDEV/global/Jiarui.Dong/JEDI/GlobalWorkflow/para_gfs/glopara_dump"`
- `land_analysis.py` introduces a method `prepare_IMS` for this type of data.
- Updates are necessary in the `GDASApp` repo.  See companion PR https://github.com/NOAA-EMC/GDASApp/pull/472

This job only runs at the `18z` cycle in the workflow.  This is controlled in the XML and not in the script.
Consequently, updates are in the `workflow/` scripts to make this happen.

Other updates included:
- Adds a method (and associated test) to provide a datetime in the form of Julian day of the year.  The method is `to_julian` in `timetools.py`
- `executable.py` was updated to accept a `list` of arguments (in addition to `str`).

**Note:**
- `calcfIMS.exe` is a serial executable and yet requires to be run with `srun`.  I recommend compiling this executable without MPI. @jiaruidong2017 @CoryMartin-NOAA 
- There are several deprecation warnings from the python IODA converter that will need to be addressed.  I recommend creating an issue to be resolved later. @jiaruidong2017 @CoryMartin-NOAA 

Part of Land DA work for snow depth assimilation, for which there is no issue open.

**Type of change**

- [X] New feature (non-breaking change which adds functionality)
- [X] This change requires a documentation update

**How Has This Been Tested?**
A standalone test was created on Hera at:
`/scratch1/NCEPDEV/stmp2/Rahul.Mahajan/EXPDIR/landda`
See `preplandobs.log` for details.
  
**Checklist**

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] New and existing tests pass with my changes
- [X] Any dependent changes have been merged and published

### CODE REVIEW

File: sorc/checkout.sh [hunk 1]
```diff
@@ -165,7 +165,7 @@ if [[ ${checkout_gsi} == "YES" ]]; then
 fi
 
 if [[ ${checkout_gdas} == "YES" ]]; then
-  checkout "gdas.cd" "https://github.com/NOAA-EMC/GDASApp.git" "aaf7caa"; errs=$((errs + $?))
+  checkout "gdas.cd" "https://github.com/NOAA-EMC/GDASApp.git" "feature/prepland"; errs=$((errs + $?))
```
Comment (CoryMartin-NOAA) [inline sorc/checkout.sh]: Note: to update this to whatever the stable nightly hash will be tomorrow AM.

File: ush/python/pygfs/task/land_analysis.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+        self.task_config = AttrDict(**self.config, **self.runtime_config, **local_dict)
+
+    @logit(logger)
+    def prepare_IMS(self: Analysis) -> None:
+        """Prepare the IMS data for a global land analysis
+
+        This method will prepare IMS data for a global land analysis using JEDI.
+        This includes:
+        - staging model backgrounds
+        - generating a card file for IMS processing
```
Comment (CoryMartin-NOAA) [inline ush/python/pygfs/task/land_analysis.py]: what is a card file?

File: workflow/rocoto/workflow_xml.py [hunk 3]
```diff
@@ -118,6 +118,12 @@ def _get_cycledefs_cycled(self):
         sdate = sdate + to_timedelta(interval)
         strings.append(f'\t<cycledef group="gdas">{sdate.strftime("%Y%m%d%H%M")} {edate.strftime("%Y%m%d%H%M")} {interval}</cycledef>')
 
+        if self._app_config.do_jedilandda:
+            sdate_land = sdate.replace(hour=18, minute=0, second=0)
+            edate_land = edate
+            if edate >= sdate:
+                strings.append(f'\t<cycledef group="gdas_land_prep">{sdate_land.strftime("%Y%m%d%H%M")} {edate_land.strftime("%Y%m%d%H%M")} 24</cycledef>')
```
Comment (CoryMartin-NOAA) [inline workflow/rocoto/workflow_xml.py]: Workflow Q for you. So landanlinit, run, final will run every cycle, the prep only 18z (for the IMS). Where does the different dependency depending on cycle come into play in the XML generation?

File: ush/python/pygfs/task/land_analysis.py (hunk 2 above)
Comment (aerorahul) [reply ush/python/pygfs/task/land_analysis.py]: No idea. I assume it is the output from `calcfIMS.exe`.  @jiaruidong2017 would know for sure.

File: workflow/rocoto/workflow_xml.py (hunk 3 above)
Comment (aerorahul) [reply workflow/rocoto/workflow_xml.py]: I haven't reviewed the `landanlinit` and `landanlrun` scripts yet on how they use the observation data created from `preplandobs`.  Once I see what that is, I may be able to add a dependency of `preplandobs` on the 18z cycle of `landanlinit` only.
I don't have an answer until I can review the actual use of this observation.

File: ush/python/pygfs/task/land_analysis.py (hunk 2 above)
Comment (CoryMartin-NOAA) [reply ush/python/pygfs/task/land_analysis.py]: I think this probably means `generating a namelist for IMS processing` ?

File: sorc/checkout.sh (hunk 1 above)
Comment (aerorahul) [reply sorc/checkout.sh]: will update to the hash on Monday.

File: ush/python/pygfs/task/land_analysis.py (hunk 2 above)
Comment (aerorahul) [reply ush/python/pygfs/task/land_analysis.py]: possible.  I'll update the comment on Monday with the hash when more reviews come in.

File: workflow/rocoto/workflow_xml.py (hunk 3 above)
Comment (CoryMartin-NOAA) [reply workflow/rocoto/workflow_xml.py]: got it, okay we will hold for now

File: ush/python/pygfs/task/land_analysis.py (hunk 2 above)
Comment (jiaruidong2017) [reply ush/python/pygfs/task/land_analysis.py]: Yes, it is namelist for IMS processing

**Review by jiaruidong2017 — CHANGES_REQUESTED** [review]
@aerorahul First thank you very much for your great efforts. The `prepare_IMS` in the `land_analysis.py` is so nice and so pretty to me, and I have learned a lot from these codes. I cloned your branch and compiled it. When I tried to setup the experiment and workflow, it is as below: 

image

I didn't find the `preplandobs` job there. Did I miss something?

**Comment by aerorahul** [comment]
@jiaruidong2017 
you should see that job only in the 18z cycle.

**Comment by jiaruidong2017** [comment]
> @jiaruidong2017 you should see that job only in the 18z cycle.

Thanks @aerorahul 
But it seems the `landanlinit` job wont start because there are not `preplandobs` job available at other cycles. Do we need to modify the `landanlinit` dependencies in the workflow_tasks.py to start the `landanlinit` job after `prep` in the other cycles?

**Comment by jiaruidong2017** [comment]
@aerorahul I merged other land DA jobs (`landanlinit`, `landanlrun` and `landanlfinal`) into your `feature/prepland` branch, and I conducted the test runs as below: 

image
image
image
image
image
image
image

See the job monitoring processes above, the `preplandobs` job runs successful at the `18z` cycle. However, the workflow seemed weirds after the `18z` cycle, and they were `202103231801`, `202103231802`, and `202103231803` for `preplandobs` jobs. Do you know what these `preplandobs` processes are? Thanks.

**Comment by WalterKolczynski-NOAA** [comment]
@jiaruidong2017 Can you point me to the rocoto XML?

**Comment by jiaruidong2017** [comment]
@WalterKolczynski-NOAA it is /scratch1/NCEPDEV/global/Jiarui.Dong/JEDI/GlobalWorkflow/para_gfs/landid8888/landid8888.xml on hera. Thanks.

File: workflow/rocoto/workflow_xml.py (hunk 3 above)
Comment (WalterKolczynski-NOAA) [inline workflow/rocoto/workflow_xml.py]:
```suggestion
                strings.append(f'\t<cycledef group="gdas_land_prep">{sdate_land.strftime("%Y%m%d%H%M")} {edate_land.strftime("%Y%m%d%H%M")} 24:00:00</cycledef>')
```

Comment (WalterKolczynski-NOAA) [reply workflow/rocoto/workflow_xml.py]: I think this is the problem: the interval in the cycledef doesn't have the zeros for minutes and seconds. Compare to the others in the XML. So there is a prep land obs job for almost every minute (I'm not sure how '24' translates into 'most minutes'.)

**Review by jiaruidong2017 — APPROVED** [review]
Thanks again for your efforts.

File: workflow/rocoto/workflow_tasks.py [hunk 4]
```diff
@@ -510,8 +510,23 @@ def aeroanlfinal(self):
...
+
+        resources = self.get_resource('preplandobs')
+        task = create_wf_task('preplandobs', resources, cdump=self.cdump, envar=self.envars, dependency=dependencies,
+                              cycledef=f'{self.cdump}_land_prep')
+
+        return task
+
     def landanlinit(self):
 
+        # TODO: Q: what from gdaspost -6H or atmf009.nc does landanlinit depend on?
```
Comment (aerorahul) [inline workflow/rocoto/workflow_tasks.py]: @jiaruidong2017 
Can you please answer this question?
Please take a look at the next 2 dependencies that were added to this job.

Comment (jiaruidong2017) [reply workflow/rocoto/workflow_tasks.py]: @aerorahul I just followed the settings from `aeroanlinit`. I think the `landanlinit` may just require model background from `prep`. @CoryMartin-NOAA Do you have any suggestions? Thanks.

Comment (CoryMartin-NOAA) [reply workflow/rocoto/workflow_tasks.py]: @jiaruidong2017 I believe it depends on the 6 hour RESTART files, no? which I guess would be the atmf009 for simplicity sake

Comment (aerorahul) [reply workflow/rocoto/workflow_tasks.py]: Ok. Thanks.
Completion of `gdasprep` should be enough to capture that, instead of the data dependency on `atmf009.nc` from previous cycle.
I will update the dependencies.

I assume this will also be true for the `aeroanlinit` @CoryMartin-NOAA?

Comment (jiaruidong2017) [reply workflow/rocoto/workflow_tasks.py]: @CoryMartin-NOAA Yes, it depends on the 6-hr RESTART files.

Comment (CoryMartin-NOAA) [reply workflow/rocoto/workflow_tasks.py]: @aerorahul yes gdasprep should be all aeroanlinit needs

Comment (aerorahul) [reply workflow/rocoto/workflow_tasks.py]: I have resolved and simplified the dependencies for both `landanlinit` and `aeroanlinit`.

File: workflow/rocoto/workflow_xml.py (hunk 3 above)
Comment (aerorahul) [reply workflow/rocoto/workflow_xml.py]: I think we resolved this.

File: workflow/rocoto/workflow_tasks.py [hunk 5]
```diff
@@ -466,21 +466,9 @@ def atmanlfinal(self):
...
-        dump_path = self._template_to_rocoto_cycstring(self._base["COM_OBSDMP_TMPL"],
-                                                       {'DMPDIR': dmpdir, 'DUMP_SUFFIX': dump_suffix})
-
         deps = []
-        dep_dict = {'type': 'metatask', 'name': 'gdaspost', 'offset': '-06:00:00'}
-        deps.append(rocoto.add_dependency(dep_dict))
-        data = f'{atm_hist_path}/gdas.t@Hz.atmf009.nc'
-        dep_dict = {'type': 'data', 'data': data, 'offset': '-06:00:00'}
-        deps.append(rocoto.add_dependency(dep_dict))
         dep_dict = {'type': 'task', 'name': f'{self.cdump}prep'}
```
Comment (WalterKolczynski-NOAA) [inline workflow/rocoto/workflow_tasks.py:470]: Why the changes to the aerosol analysis dependencies?

**Review by WalterKolczynski-NOAA — CHANGES_REQUESTED** [review]
Looks really good. Question about aerosol dependencies, and still need a hash for GDASapp.

File: workflow/rocoto/workflow_tasks.py (hunk 5 above)
Comment (aerorahul) [reply workflow/rocoto/workflow_tasks.py:470]: because they are not needed. And I am breaking my own rule of not adding unrelated code updates.
Do you want me to pull these out in its own PR?

Comment (WalterKolczynski-NOAA) [reply workflow/rocoto/workflow_tasks.py:470]: I'm fine leaving the changes as long as we are sure the dependencies are not needed. I guess `prep` already has the same ones?

Comment (CoryMartin-NOAA) [reply workflow/rocoto/workflow_tasks.py:470]: yeah the dependencies are 1) the previous forecast completing and 2) observations being available. `prep` should take care of those I think

File: sorc/checkout.sh (hunk 1 above)
Comment (aerorahul) [reply sorc/checkout.sh]: GDASApp hash has been updated.

**Comment by aerorahul** [comment]
> Looks really good. Question about aerosol dependencies, and still need a hash for GDASapp.

@WalterKolczynski-NOAA GDASapp hash has been updated.

**Review by CoryMartin-NOAA — APPROVED** [review]
Thanks @aerorahul for working on this!

**Review by WalterKolczynski-NOAA — APPROVED** [review]
